pkg/watermark/fetch/edge_fetcher_test.go

```diff
@@ -1342,3 +1363,400 @@ func TestFetcherWithSameOTBucket(t *testing.T) {
+		}
+		err = hbStore.DeleteKey(ctx, "p1")
+		assert.NoError(t, err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var err error
+		// run p2 for 20 seconds
+		for i := 0; i < 20; i++ {
+			err = hbStore.PutKV(ctx, "p2", []byte(fmt.Sprintf("%d", time.Now().Unix())))
+			assert.NoError(t, err)
+			time.Sleep(1 * time.Second)
+		}
+	}()
+
+	allProcessors := processorManager.GetAllProcessors()
+	for len(allProcessors) != 2 {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				t.Fatalf("expected 2 processors, got %d: %s", len(allProcessors), ctx.Err())
+			}
+		default:
+			time.Sleep(1 * time.Millisecond)
+			allProcessors = processorManager.GetAllProcessors()
+		}
+	}
+
+	for allProcessors["p1"].GetOffsetTimelines()[0].Dump() != "[1651161600300:102] -> [1651161600200:101] -> [1651161600100:100] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" &&
+		allProcessors["p1"].GetOffsetTimelines()[1].Dump() != "[1651161600300:102] -> [1651161600200:101] -> [1651161600100:100] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" &&
+		allProcessors["p1"].GetOffsetTimelines()[2].Dump() != "[1651161600300:102] -> [1651161600200:101] -> [1651161600100:100] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				t.Fatalf("expected p1 has the offset timelines [1651161600300:102] -> [1651161600200:101] -> [1651161600100:100]..., [1651161600300:102] -> [1651161600200:101] -> [1651161600100:100]..., "+
+					"[1651161600300:102] -> [1651161600200:101] -> [1651161600100:100]... got %s %s %s: %s", allProcessors["p1"].GetOffsetTimelines()[0].Dump(), allProcessors["p1"].GetOffsetTimelines()[1].Dump(), allProcessors["p1"].GetOffsetTimelines()[2].Dump(), ctx.Err())
+			}
+		default:
+			time.Sleep(1 * time.Millisecond)
+			allProcessors = processorManager.GetAllProcessors()
+		}
+	}
+
+	assert.True(t, allProcessors["p1"].IsActive())
+	assert.True(t, allProcessors["p2"].IsActive())
+
+	// "p1" is deleted after 5 loops
+	for !allProcessors["p1"].IsDeleted() {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				t.Fatalf("expected p1 to be deleted: %s", ctx.Err())
+			}
+		default:
+			time.Sleep(1 * time.Millisecond)
+			allProcessors = processorManager.GetAllProcessors()
+		}
+	}
+
+	allProcessors = processorManager.GetAllProcessors()
+	assert.Equal(t, 2, len(allProcessors))
+	assert.True(t, allProcessors["p1"].IsDeleted())
+	assert.True(t, allProcessors["p2"].IsActive())
+
+	_ = fetcher.GetWatermark(isb.SimpleStringOffset(func() string { return strconv.FormatInt(testOffset, 10) }), 0)
+	_ = fetcher.GetWatermark(isb.SimpleStringOffset(func() string { return strconv.FormatInt(testOffset, 10) }), 1)
+	_ = fetcher.GetWatermark(isb.SimpleStringOffset(func() string { return strconv.FormatInt(testOffset, 10) }), 2)
```

Comment:
Should we verify the return value of GetWatermark? If not, what are we checking here?

Reply:
we are deleting the processor using the GetWatermark method, if you look at the GetWatermark implementation in some cases we are deleting the processor. So we are testing those cases here.

Reply:
Got it. L1514-1521 verified that if inputOffset is less than head offset of a processor, the processor won't get deleted.